### PR TITLE
Improve categorical renderer style

### PR DESCRIPTION
**Pull request by thinkh**
Closes Caleydo/lineupjs#487
Closes Caleydo/lineupjs#489

image

### CODE REVIEW

**Comment by thinkh** [comment]
I'll try to add this issue: https://github.com/Caleydo/lineupjs/issues/489

**Comment by thinkh** [comment]
I also integrated issue #489, which looks like this now:

image

The implementation calculates the lightness of the background-color and applies either black or white as text color. For browsers that doesn't support CSS custom properties we use a black text color as fallback (see screenshot in the first comment).

File: src/utils.ts [hunk 1]
```diff
@@ -576,3 +576,39 @@ export function deriveColors(columns: IColumnDesc[]) {
...
+	return Math.round((Math.min(...rgbColor) + Math.max(...rgbColor))/2.55/2);
+}
+
+/**
+ * Adapts the text color for a given background color (as HEX value)
+ * @param {string} bgHexColor as `#ff0000`
+ * @returns {string} returns `black` or `white` for best contrast
+ */
+export function adaptTextColorToBgColor(bgHexColor:string):string {
+  return colorLightness(hexToRgb(bgHexColor)) > 50 ? 'black' : 'white';
```
Comment (sgratzl) [inline src/utils.ts]: why not using d3 for that -> `d3.hsl(rgb)...`

File: src/styles/renderer/_catdistributionbar.scss [hunk 2]
```diff
@@ -10,5 +10,17 @@
...
+    color: transparent;
+
+    &:before {
+      content: attr(data-cat);
+      position: absolute;
+      left: 5px;
+      right: 3px;
+      overflow: hidden;
+      text-overflow: ellipsis;
+      color: var(--text-color, black); // custom property `--text-color` is set in CategoricalStackedDistributionlRenderer.ts
```
Comment (sgratzl) [inline src/styles/renderer/_catdistributionbar.scss]: does every browser (firefox esr, chrome, edge) support that?

File: src/renderer/CategoricalStackedDistributionlRenderer.ts [hunk 3]
```diff
@@ -22,7 +22,7 @@ export default class CategoricalStackedDistributionlRenderer implements ICellRen
   createGroupDOM(col: ICategoricalColumn & Column): IDOMGroupRenderer {
     const colors = col.categoryColors;
     const labels = col.categoryLabels;
-    const bins = col.categories.map((c, i) => `<div style="background-color: ${colors[i]}" title="${labels[i]}: 0" data-cat="${c}">${labels[i]}</div>`).join('');
+    const bins = col.categories.map((c, i) => `<div style="background-color: ${colors[i]}; --text-color: ${adaptTextColorToBgColor(colors[i])}" title="${labels[i]}: 0" data-cat="${c}">${labels[i]}</div>`).join('');
```
Comment (sgratzl) [inline src/renderer/CategoricalStackedDistributionlRenderer.ts]: isn't the labels already part of the content -> and later on the `data-cat` will be overlayed?

Comment (thinkh) [reply src/renderer/CategoricalStackedDistributionlRenderer.ts]: Yes, the label is available as element content. However, as you found out it's not possible to shift it. That's why I made the element text invisible with `color: transparent` and set the color for the pseudo-element using this CSS custom property (or black as default).

File: src/styles/renderer/_catdistributionbar.scss (hunk 2 above)
Comment (thinkh) [reply src/styles/renderer/_catdistributionbar.scss]: Yes, every new browser: https://caniuse.com/#feat=css-variables

File: src/utils.ts (hunk 1 above)
Comment (thinkh) [reply src/utils.ts]: I'll check this.

File: src/styles/renderer/_catdistributionbar.scss (hunk 2 above)
Comment (sgratzl) [reply src/styles/renderer/_catdistributionbar.scss]: what happens it is isn't supported, such as in in Firefox ESR 52...

File: src/renderer/CategoricalStackedDistributionlRenderer.ts (hunk 3 above)
Comment (sgratzl) [reply src/renderer/CategoricalStackedDistributionlRenderer.ts]: why do you add then the label at all, if you hide it anyhow?

Comment (thinkh) [reply src/renderer/CategoricalStackedDistributionlRenderer.ts]: Right. I changed it now and use a `span` element instead, which makes the pseudo-element and CSS custom properties obsolete.

File: src/styles/renderer/_catdistributionbar.scss (hunk 2 above)
Comment (thinkh) [reply src/styles/renderer/_catdistributionbar.scss]: The solution is to use `@supports` as you can see here: https://github.com/Caleydo/lineupjs/pull/506/commits/54811aa1aeb436656cdc404edf201724e00cda72

**Comment by thinkh** [comment]
@sgratzl I addressed all your comments. Please have a look again.

**Pull request merged**